### PR TITLE
DataCollatorForPreference checking 'margin' in all examples

### LINKED ISSUE #5539: fix: DataCollatorForRewardModelingDataset checks margin only on examples[0], causing silent data loss after shuffle
## Problem

In `trl/trainer/reward_trainer.py`, `DataCollatorForRewardModelingDataset.torch_call()` checks for the `"margin"` key only on `examples[0]`:

```python
if "margin" in examples[0]:
    margins = torch.tensor([example["margin"] for example in examples], dtype=torch.float)
```

This has two failure modes:

1. **Silent data loss**: After dataset shuffling, if the first example lacks `"margin"` but subsequent examples have it, margins are silently dropped and the margin-aware reward loss is skipped with no warning.
2. **Runtime `KeyError`**: If the first example has `"margin"` but a later one does not (inconsistent dataset), the list comprehension crashes.

## Expected Behavior

All examples should be checked for the `"margin"` key. Missing margins should default to `0.0` gracefully.

## Fix

```python
if any("margin" in example for example in examples):
    margins = torch.tensor([example.get("margin", 0.0) for example in examples], dtype=torch.float)
```

### PR DESCRIPTION
**Pull request by antoinsader**
# What does this PR do?

Fixes #5539 
In trl/trainer/reward_trainer.py, DataCollatorForRewardModelingDataset.torch_call() checks for the "margin" key only on examples[0]:

if "margin" in examples[0]:
    margins = torch.tensor([example["margin"] for example in examples], dtype=torch.float)

It was fixed to use `any('margin' in example for example in examples)` to check if the examples has margin. 

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. https://github.com/huggingface/trl/issues/5539
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [x] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Low risk behavior change confined to data collation: optional fields are now detected across the whole batch and missing values default to 0.0, reducing KeyError risk for mixed-schema datasets.
> 
> **Overview**
> Fixes preference-data collation when optional fields are present only in some batch elements.
> 
> `DataCollatorForPreference` in `reward_trainer.py` now detects `margin` across all examples and fills missing margins with `0.0`; `dpo_trainer.py` does the same for `ref_chosen_logps`/`ref_rejected_logps`. New tests cover mixed batches where these optional keys are missing on some examples.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c77e4a27fd7915e57e7ceaf09cfac7474a040c24. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
https://github.com/huggingface/trl/issues/5539#issuecomment-4255437708

If not I don't think there anything to "fix"
